File: python/code_injection/eval_usage.py
from flask import Flask, request

app = Flask(__name__)


# VULN 1: eval() on user-supplied discount formula - arbitrary code execution
@app.route("/cart/apply-discount", methods=["POST"])
def apply_discount():
    formula = request.json.get("formula")  # e.g., "price * 0.9"
    price = float(request.json.get("price", 0))
    result = eval(formula)
    return {"discounted_price": result}


# VULN 2: exec() on user-controlled shipping rule script
@app.route("/admin/shipping-rules", methods=["POST"])
def update_shipping_rules():
    rule_code = request.json.get("rule")
    exec(rule_code)
    return {"status": "rules updated"}


# VULN 3: eval() in product filter expression - user controls filter logic
@app.route("/products/filter")
def filter_products():
    filter_expr = request.args.get("filter")
    products = get_all_products()
    filtered = [p for p in products if eval(filter_expr, {"p": p})]
    return {"products": filtered}


def get_all_products():
    return []

File: ruby/code_injection/eval_usage.rb
require 'sinatra'
require 'active_record'

# VULN 1: eval() on user-supplied formula - arbitrary Ruby code execution for discount calculation
post '/cart/apply_discount' do
  formula = params[:formula]
  cart_total = session[:cart_total].to_f
  # eval executes attacker-controlled Ruby: e.g. formula="`cat /etc/passwd`"
  discount = eval(formula)
  session[:discount] = discount
  { discount: discount, new_total: cart_total - discount }.to_json
end

# VULN 2: instance_eval on user-supplied shipping rule string - code execution in current object context
post '/shipping/calculate' do
  rule = params[:rule]
  order = Order.find(params[:order_id])
  # instance_eval gives attacker access to order object methods and full Ruby
  shipping_cost = order.instance_eval(rule)
  { shipping_cost: shipping_cost }.to_json
end

# VULN 3: Kernel.eval on user-supplied product filter - arbitrary code execution for product listing
get '/products/filtered' do
  filter = params[:filter]
  products = Product.all.to_a
  # Kernel.eval executes attacker-supplied expression with full access to Ruby runtime
  result = Kernel.eval(filter)
  { products: result }.to_json
end
